File: pat/1155.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isMax = true, isMin = true;

int *heap;
int n;
vector<int> path;

void show() {
    for (int i = 0; i < path.size(); i ++) {
        if (i != 0) {
            cout << " ";
        }
        cout << path[i];
    }
    cout << endl;
}

void travel(int index) {
    if (index >= n) {
        return;
    }
    if (2 * index + 1 >= n) {
        path.push_back(heap[index]);
        show();
        path.pop_back();
    } else {
        path.push_back(heap[index]);
        if (2 * index + 2 < n) {
            travel(2 * index + 2);
            travel(2 * index + 1);
            if (heap[2 * index + 1] < heap[index] || heap[2 * index + 2] < heap[index]) {
                isMin = false;
            }
            if (heap[2 * index + 1] > heap[index] || heap[2 * index + 2] > heap[index]) {
                isMax = false;
            }
        } else if (2 *index + 2 == n) {
            travel(2 * index + 1);
            if (heap[2 * index + 1] < heap[index]) {
                isMin = false;
            }
            if (heap[2 * index + 1] > heap[index]) {
                isMax = false;
            }
        }
        path.pop_back();
    }

}

int main()
{
    cin >> n;
    heap = new int[n];
    for (int i = 0; i < n; i ++) {
        cin >> heap[i];
    }
    travel(0);
    if (!isMax && !isMin) {
        cout << "Not Heap" << endl;
    } else if (isMax) {
        cout << "Max Heap" << endl;
    } else if (isMin){
        cout << "Min Heap" << endl;
    }
    return 0;
}

File: pat-go/1155.go
package main

import "fmt"

var a = make([]int, 0)
var is_max, is_min = true, true
var path = make([]int, 0)

// 打印一条路径的结果
func show() {
	for i := 0; i < len(path); i++ {
		if i != 0 {
			fmt.Printf(" ")
		}
		fmt.Printf("%d", path[i])
	}
	fmt.Printf("\n")
}

func travel(index int) {
	if index >= len(a) {
		return
	}
	path = append(path, a[index])
	// 超出了堆数组，说明到底部了
	if 2*index+1 >= len(a) {
		show()
		path = path[0 : len(path)-1]
	} else {
		var left = 2*index + 1
		var right = 2*index + 2
		if right < len(a) {
			travel(right)
			travel(left)
			// 判断是否出现了不满足最大/最小堆的情况
			if a[left] < a[index] || a[right] < a[index] {
				is_min = false
			}
			if a[left] > a[index] || a[right] > a[index] {
				is_max = false
			}
		} else if right == len(a) {
			travel(left)
			if a[left] < a[index] {
				is_min = false
			}
			if a[left] > a[index] {
				is_max = false
			}
		}
		path = path[0 : len(path)-1]
	}
}

func main() {
	var n int
	fmt.Scanf("%d", &n)
	a = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%d", &a[i])
	}
	travel(0)
	if is_max && !is_min {
		fmt.Println("Max Heap")
	} else if is_min && !is_max {
		fmt.Println("Min Heap")
	} else {
		fmt.Println("Not Heap")
	}
}
